packages/graph/hash_graph/lib/graph/src/types/schema/property_type.rs

```diff
@@ -0,0 +1,727 @@
+#[serde(
+    try_from = "PropertyTypeObjectRepr",
+    tag = "type",
+    rename = "object",
+    rename_all = "camelCase"
+)]
+pub struct PropertyTypeObject {
+    /// Property names must be a valid URI to a property-type
+    properties: HashMap<Uri, ValueOrArray<PropertyTypeReference>>,
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    required: Vec<Uri>,
+}
+
+impl PropertyTypeObject {
+    /// Creates a new `PropertyTypeObject` without validating.
+    #[must_use]
+    pub const fn new_unchecked(
+        properties: HashMap<Uri, ValueOrArray<PropertyTypeReference>>,
+        required: Vec<Uri>,
+    ) -> Self {
+        Self {
+            properties,
+            required,
+        }
+    }
+
+    /// Creates a new `PropertyTypeObject` with the given properties and required properties.
+    ///
+    /// # Errors
+    ///
+    /// - [`ValidationError::PropertyMissing`] if a required property is not a key in `properties`.
+    pub fn new(
+        properties: HashMap<Uri, ValueOrArray<PropertyTypeReference>>,
+        required: Vec<Uri>,
+    ) -> Result<Self, ValidationError> {
+        let object = Self::new_unchecked(properties, required);
+        object.validate()?;
+        Ok(object)
+    }
+}
+
+impl Validate for PropertyTypeObject {
+    fn validate(&self) -> Result<(), ValidationError> {
+        for uri in &self.required {
+            if !self.properties.contains_key(uri) {
+                return Err(ValidationError::PropertyMissing(uri.clone()));
+            }
+        }
+        Ok(())
+    }
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+enum ObjectTypeTag {
+    Object,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+struct PropertyTypeObjectRepr {
+    #[serde(rename = "type")]
+    _type: ObjectTypeTag,
+    /// Property names must be a valid URI to a property-type
+    properties: HashMap<Uri, ValueOrArray<PropertyTypeReference>>,
+    #[serde(default)]
+    required: Vec<Uri>,
+}
+
+impl TryFrom<PropertyTypeObjectRepr> for PropertyTypeObject {
+    type Error = ValidationError;
+
+    fn try_from(object: PropertyTypeObjectRepr) -> Result<Self, ValidationError> {
+        Self::new(object.properties, object.required)
+    }
+}
+
+mod values {
+    #![allow(
+        clippy::use_self,
+        clippy::wildcard_imports,
+        reason = "Weird false positive on `ArrayOfPropertyValues` which somehow can't be disabled \
+                  locally"
+    )]
+
+    use super::*;
+
+    #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+    #[serde(untagged)]
+    pub enum PropertyValues {
+        PropertyTypeObject(PropertyTypeObject),
+        DataTypeReference(DataTypeReference),
+        ArrayOfPropertyValues(Array<OneOf<Self>>),
+    }
+}
+pub use self::values::PropertyValues;
+
+impl PropertyValues {
+    #[must_use]
+    fn datatype_references(&self) -> Vec<&DataTypeReference> {
```

Comment:
should this be called `data_type_references` as it is on the `PropertyType` impl?

Reply:
Applied: fd63290cd1bfe78ecea23e5e3c0c23d9895cb479